clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/cache/CatsOnDemandCacheUpdater.groovy

```diff
@@ -57,13 +61,27 @@ class CatsOnDemandCacheUpdater implements OnDemandCacheUpdater {
   OnDemandCacheUpdater.OnDemandCacheStatus handle(OnDemandAgent.OnDemandType type, Collection<OnDemandAgent> onDemandAgents, Map<String, ? extends Object> data) {
     boolean hasOnDemandResults = false
     for (OnDemandAgent agent : onDemandAgents) {
+      boolean atomicScheduler = false;
+      AgentScheduler scheduler = null
+      if (agent instanceof AgentSchedulerAware) {
+        scheduler = agent.agentScheduler
+        atomicScheduler = scheduler.atomic
+      }
       try {
+        AgentLock lock = null;
+        if (atomicScheduler && !(lock = scheduler.tryLock((Agent) agent))) {
+          hasOnDemandResults = true // force Orca to retry
+          continue;
+        }
         final long startTime = System.nanoTime()
         def providerCache = catsModule.getProviderRegistry().getProviderCache(agent.providerName)
         OnDemandAgent.OnDemandResult result = agent.handle(providerCache, data)
         if (result) {
+          if (atomicScheduler && !(scheduler.tryRelease(lock))) {
+            hasOnDemandResults = true // force Orca to retry
+          }
           if (result.cacheResult) {
-            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty()
+            hasOnDemandResults = !(result.cacheResult.cacheResults ?: [:]).values().flatten().isEmpty() && !atomicScheduler
             agent.metricsSupport.cacheWrite {
```

Comment:
this releases the lock before write the results. if that triggers the normal caching run for the agent, it may not see the on demand node when it runs (in our case for AWS we will still have to inspect that during our normal caching cycle to make sure our polled data is at least as new as our onDemand data)

I think I'd err on the side of doing release after write 

also probably want to stick the tryRelease call in a finally block to make sure we release if some exception happens during handle